### PR TITLE
EMSUSD-751 - UsdUfe: Move Ufe::Attributes interface implementation

### PR DESCRIPTION
**Pull request by seando-adsk**
#### EMSUSD-751 - UsdUfe: Move Ufe::Attributes interface implementation
* Tokens moved: generic and some of the meta data.
* Functions moved:
  * prettifyName (python wrapper)
  * isConnected()
  * canRemoveSrcProperty() / canRemoveDstProperty()
  * usdTypeToUfe() / ufeTypeToUsd()
  * usdAttrFromUfeAttr() / attrFromUfeAttrInfo()
  * vtValueFromString()
  * splitString()
* Classes moved:
  * UsdAttributesHandler/UsdAttributes/UsdAttribute/UsdAttributeHolder
  * UsdShaderAttributeDef/UsdShaderAttributeHolder/
  * UsdAddAttributeCommand/UsdRemoveAttributeCommand/UsdRenameAttributeCommand
* New (optional) DCC message (info/warning/error) functions.
* Removed Maya setting isAttributeLocked function as a default implementation exists in UsdUfe.
* Moved code from UsdShaderNodeDefHandler::usdDefinition() to UsdUfe::usdShaderNodeFromSceneItem().
* Removed mayaUsd inline prettifyName().

#### EMSUSD-751 - UsdUfe: Move Ufe::Attributes interface implementation
* Removed mayaUsd inline SanitizeName().

#### EMSUSD-751 - UsdUfe: Move Ufe::Attributes interface implementation
* Moved 'using namespace UsdUfe' to mayaUsd to make sure that UsdUfe is properly using namespace.

### CODE REVIEW

File: lib/usdUfe/ufe/Global.h [hunk 1]
```diff
@@ -48,6 +49,7 @@ struct USDUFE_PUBLIC DCCFunctions
     SaveStageLoadRulesFn saveStageLoadRulesFn = nullptr;
     IsRootChildFn        isRootChildFn = nullptr;
     UniqueChildNameFn    uniqueChildNameFn = nullptr;
+    DisplayMessageFn     displayMessageFn[static_cast<int>(MessageType::nbTypes)] = { nullptr };
```
Comment (seando-adsk) [inline lib/usdUfe/ufe/Global.h:52]: New array of (optional) DCC message functions.

File: lib/usdUfe/ufe/Utils.cpp [hunk 2]
```diff
@@ -224,7 +239,11 @@ void setIsAttributeLockedFn(IsAttributeLockedFn fn)
 
 bool isAttributedLocked(const PXR_NS::UsdAttribute& attr, std::string* errMsg /*= nullptr*/)
 {
-    return gIsAttributeLockedFn ? gIsAttributeLockedFn(attr, errMsg) : false;
+    // If we have (optional) attribute is locked function, use it.
+    // Otherwise use the default one supplied by UsdUfe.
+    if (gIsAttributeLockedFn)
+        return gIsAttributeLockedFn(attr, errMsg);
+    return Editability::isAttributeLocked(attr, errMsg);
```
Comment (seando-adsk) [inline lib/usdUfe/ufe/Utils.cpp:246]: Now that isAttributeLocked() was moved to UsdUfe, we have a default function to call.

File: lib/mayaUsd/ufe/Global.cpp [hunk 3]
```diff
@@ -175,9 +177,11 @@ MStatus initialize()
     dccFunctions.stagePathAccessorFn = MayaUsd::ufe::stagePath;
     dccFunctions.ufePathToPrimFn = MayaUsd::ufe::ufePathToPrim;
     dccFunctions.timeAccessorFn = MayaUsd::ufe::getTime;
-    dccFunctions.isAttributeLockedFn = MayaUsd::Editability::isAttributeLocked;
     dccFunctions.saveStageLoadRulesFn = MayaUsd::MayaUsdProxyShapeStageExtraData::saveLoadRules;
     dccFunctions.uniqueChildNameFn = MayaUsd::ufe::uniqueChildNameMayaStandard;
+    dccFunctions.displayMessageFn[static_cast<int>(MessageType::kInfo)] = displayInfoMessage;
+    dccFunctions.displayMessageFn[static_cast<int>(MessageType::kWarning)] = displayWarningMessage;
+    dccFunctions.displayMessageFn[static_cast<int>(MessageType::KError)] = displayErrorMessage;
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/Global.cpp:184]: Set the Maya message logging functions.

**Review by seando-adsk — COMMENTED** [review]
@deboisj / @pierrebai-adsk I'll assign both of you as code reviewers. No rush to review. I'm off for the next of this week.

File: lib/mayaUsd/ufe/Global.cpp [hunk 4]
```diff
@@ -120,6 +118,10 @@ void mayaStopWaitCursor() { MGlobal::executeCommand("waitCursor -state 0"); }
 // Note: MayaUsd::ufe::getStage takes two parameters, so wrap it in a function taking only one.
 PXR_NS::UsdStageWeakPtr mayaGetStage(const Ufe::Path& path) { return MayaUsd::ufe::getStage(path); }
 
+void displayInfoMessage(const std::string& msg) { MGlobal::displayInfo(msg.c_str()); }
```
Comment (deboisj) [inline lib/mayaUsd/ufe/Global.cpp:121]: is this going to the maya listener? (the thing showing the red text bottom right)

File: lib/usdUfe/ufe/Utils.cpp [hunk 5]
```diff
@@ -379,6 +398,239 @@ SdfPath uniqueChildPath(const UsdStage& stage, const SdfPath& path)
...
+    // Each of the display message functions is allowed to be null in which case
+    // a default function will be used for each.
+    for (int i = 0; i < static_cast<int>(MessageType::nbTypes); ++i) {
+        gDisplayMessageFn[i] = fns[i];
+    }
+}
+
+void displayMessage(MessageType type, const std::string& msg)
+{
+    // If we have an (optional) display message for the input type, use it.
```
Comment (deboisj) [inline lib/usdUfe/ufe/Utils.cpp:412]: 👍

File: lib/usdUfe/ufe/Utils.cpp [hunk 6]
```diff
@@ -379,6 +398,239 @@ SdfPath uniqueChildPath(const UsdStage& stage, const SdfPath& path)
...
+        switch (type) {
+        case MessageType::kInfo: TF_STATUS(msg); break;
+        case MessageType::kWarning: TF_WARN(msg); break;
+        case MessageType::KError: TF_RUNTIME_ERROR(msg); break;
+        default: break;
+        }
+    }
+}
+
+namespace {
```
Comment (deboisj) [inline lib/usdUfe/ufe/Utils.cpp:427]: unless specified i assumed all that is only moved / nothing new

File: lib/usdUfe/utils/editability.cpp [hunk 7]
```diff
@@ -33,15 +33,15 @@ bool isLocked(PXR_NS::UsdProperty property)
...
 
-    if (lock == MayaUsdTokens->Off) {
+    if (lock == GenericTokens->Off) {
         return false;
-    } else if (lock == MayaUsdTokens->On) {
+    } else if (lock == GenericTokens->On) {
         return true;
     } else {
-        TF_WARN("Invalid token value [%s] for maya lock will be treated as [off].", lock.data());
+        TF_WARN("Invalid token value [%s] for lock will be treated as [off].", lock.data());
```
Comment (deboisj) [inline lib/usdUfe/utils/editability.cpp:44]: isnt it still maya lock for now?

**Review by deboisj — DISMISSED** [review]
Looks good, didnt have much to say. Hopefully there is good test coverage! Thanks alot!

File: lib/mayaUsd/ufe/Global.cpp (hunk 4 above)
Comment (pierrebai-adsk) [reply lib/mayaUsd/ufe/Global.cpp:121]: Yes except being info it won't be red

File: lib/usdUfe/utils/Utils.cpp [hunk 8]
```diff
@@ -82,4 +82,20 @@ std::string prettifyName(const std::string& name)
...
 
+std::vector<std::string> splitString(const std::string& str, const std::string& separators)
+{
+    std::vector<std::string> split;
+
+    std::string::size_type lastPos = str.find_first_not_of(separators, 0);
+    std::string::size_type pos = str.find_first_of(separators, lastPos);
+
+    while (pos != std::string::npos || lastPos != std::string::npos) {
+        split.push_back(str.substr(lastPos, pos - lastPos));
```
Comment (pierrebai-adsk) [inline lib/usdUfe/utils/Utils.cpp:93]: I suppose this is moved code, but there is a bug if the string starts with separators. For example, assume the separator is ';' and the string is ";;;aaa;;;", then lastPos would be 3 (the first a) and pos would be zero (the first ;) and pos - lastPos would be 0 - 3, which when converted to unsigned would be a huge number.

Depending on the behavior of str.substr(), we either have abuffer overflow or we extract "aaa;;;" as the string.

**Review by pierrebai-adsk — DISMISSED** [review]

File: lib/mayaUsd/ufe/Global.cpp (hunk 4 above)
Comment (seando-adsk) [reply lib/mayaUsd/ufe/Global.cpp:121]: This prints to the Maya script editor and the command line. Three types of message: info, warning & error (same as USD).

**Comment by seando-adsk** [comment]
@deboisj / @pierrebai-adsk I had to merge in latest dev and then add a few build fixes (related to new Ufe method `isDefault`).

**Review by deboisj — APPROVED** [review]

**Review by pierrebai-adsk — DISMISSED** [review]

**Comment by seando-adsk** [comment]
@pierrebai-adsk Sorry had to merge in latest dev again since I merged my other clipboard PR. This should be the last time. I hope to merge this in on Thu.

**Review by pierrebai-adsk — DISMISSED** [review]

**Comment by seando-adsk** [comment]
@pierrebai-adsk Sorry one last commit to fix Linux/OSX build issues. It was because I removed the `using namespace UsdUfe` from the `api.h` file.

**Review by pierrebai-adsk — APPROVED** [review]

**Pull request merged**